src/pymor/models/basic.py

```diff
@@ -84,6 +87,60 @@ def __str__(self):
     def _compute_solution(self, mu=None, **kwargs):
         return self.operator.apply_inverse(self.rhs.as_range_array(mu), mu=mu)
 
+    def _compute_solution_d_mu_single_direction(self, parameter, index, solution, mu):
+        lhs_d_mu = self.operator.d_mu(parameter, index).apply(solution, mu=mu)
+        rhs_d_mu = self.rhs.d_mu(parameter, index).as_range_array(mu)
+        rhs = rhs_d_mu - lhs_d_mu
+        return self.operator.jacobian(solution, mu=mu).apply_inverse(rhs)
+
+    _compute_allowed_kwargs = frozenset({'use_adjoint'})
+
+    def _compute_output_d_mu(self, solution, mu, return_array=False, use_adjoint=None):
+        """compute the gradient of the output functional  w.r.t. the parameters
+
+        Parameters
+        ----------
+        solution
+            Internal model state for the given |Parameter value|
+        mu
+            |Parameter value| for which to compute the gradient
+        use_adjoint
+            if `None` use standard approach, if `True`, use
+            the adjoint solution for a more efficient way of computing the gradient.
+            See Section 1.6.2 in [HPUU09]_ for more details.
+            So far, the adjoint approach is only valid for linear models.
+
+        Returns
+        -------
+        The gradient as a numpy array.
+        """
+        if not use_adjoint:
+            return super()._compute_output_d_mu(solution, mu, return_array)
+        else:
+            assert self.output_functional is not None
+            assert self.output_functional.linear
+            dual_solutions = self.operator.range.empty()
+            for d in range(self.output_functional.range.dim):
+                dual_problem = self.with_(operator=self.operator.H, rhs=self.output_functional.H.as_range_array(mu)[d])
```

Comment:
```suggestion
            output_functional_array = self.output_functional.H.as_range_array(mu)
            dual_solutions = self.operator.range.empty()
            for d in range(self.output_functional.range.dim):
                dual_problem = self.with_(operator=self.operator.H, rhs=output_functional_array[d])
```

Although unlikely, converting to a VA might be costly.